### PR TITLE
Use simpler algorithm for updating fields in arrays.

### PR DESCRIPTION
**Pull request by yuchenshi**
WANT_LGTM=all

This is complicated logic even if I tried to simplify it. Reviewers, please make sure it is easy to follow.

### CODE REVIEW

File: src/components/Firestore/DocumentPreview/api.tsx [hunk 1]
```diff
@@ -55,104 +55,32 @@ function adjustPayloadForArray(
...
-    if (
-      topLevelArray === nestedObjectToModify &&
-      Number(lastFieldName(fieldPath)) >= (topLevelArray as {}[]).length
-    ) {
-      // Appending a new element to the top-level array
-      return [
-        new firestore.FieldPath(...topLevelKey),
-        firestore.FieldValue.arrayUnion(value),
-      ];
+  let cur: FirestoreAny | undefined = doc;
```
Comment (tjlav5) [inline src/components/Firestore/DocumentPreview/api.tsx]: When would current be undefined?

Comment (tstirrat) [reply src/components/Firestore/DocumentPreview/api.tsx]: perhaps on a missing doc when deep navigating?

Comment (tstirrat) [reply src/components/Firestore/DocumentPreview/api.tsx]: Oh sorry, yes you're right. it doesn't look like it would be undefined ever

Comment (yuchenshi) [reply src/components/Firestore/DocumentPreview/api.tsx]: `cur` will be undefined when adding a key to an object (e.g. set path `f.g` on `{"f": 1}`), with the assignment below. `doc` won't ever be `undefined` but `cur` will, thus the typing here.

**Pull request merged**